### DISCUSSION TITLE
Where was the design discussion for ObservablePropertyAttribute?

### DISCUSSION THREAD

**Discussion by symbiogenesis** [CommunityToolkit/dotnet#554] [Q&A] [answered]
I was developing Silverlight back in the day, and the tool Fody set my expectations pretty high for the ergonomics of this kind of thing.

To me, the approach that Fody chose is far more intuitive and explicit. It just makes more sense that the property should be the one to have the attribute.

What was the reason for making it based on the field? Was it a limitation of the source generator approach, was it a performance thing, or was it a conscious design decision to promote accessing the field within the ViewModel, or something else?

**Comment by Sergio0694** [comment, marked as answer]
You can find the original issue and plenty of discussions here: https://github.com/CommunityToolkit/dotnet/issues/8. The TLDR is: we'd much rather use properties, but unfortunately C# just doesn't support partial properties yet, so using fields is the best we could do. Assuming C# 12 does get partial properties (see https://github.com/dotnet/csharplang/issues/6420), I do plan to add support for using `[ObservableProperty]` on properties too, for sure. Hope this helps! 🙂

**Reply by edwardmiller-mesirow** [reply]
Congratulations on shipping this feature. Wouldn't it make sense to deprecate the field-based approach?

**Comment by symbiogenesis** [comment]
Ok, I understand. Source generators do not rewrite any code, but can only add. Fody was more magical than that, for better or worse.

It still feels like partial properties aren't optimal, and that some other undiscovered pattern exists that would not require a language modification or overly magical aspect-oriented approach.

If the [ObservableProperty] attribute were switched to use the properties, could we allow [RelayCommand] attributes to simply automatically search for all properties in the same class that have an ObservablePropertyAttribute, and call the PropertyChanged event for all code paths that would hit a setter in a control flow graph?

**Reply by symbiogenesis** [reply]
I guess that the commands would not have a deterministic control flow knowable at compile time, without rewriting code. I'm just thinking out loud. :)

**Comment by prlcutting** [comment]
I agree that allowing `[ObservableProperty]` on a property declaration would be much more intuitive. It would also allow for defining XML documentation comments too.

**Reply by Sergio0694** [reply]
It is allowed now, we shipped that feature in preview two years ago 🙂

**Reply by prlcutting** [reply]
Ha! Who knew? 😜 Thanks!

What threw me off was this [ObservableProperty attribute](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/observableproperty) documentation. It makes no reference to being able to set this on properties, only fields, and all the examples are with fields. Has the documentation not been updated, or am I looking at an old/wrong version of it or something?

Also, when I just tried it, it didn't work initially. I realized that it requires setting `<LangVersion>preview</LangVersion>` to get access to the `field` keyword functionality. Will there be an update to `CommunityToolkit.Mvvm` now that .NET 10 is released and the `field` keyword is mainstream in C#14? I'd rather not specify the preview language version because that might enable other language features that we don't want in our code base.

Thanks again.

**Reply by Sergio0694** [reply]
Yeah the update is ready to go, just need to find time to do one last validation pass and hit the publish button 😄

**Reply by prlcutting** [reply]
Awesome! I'll look forward to the release. Thanks!
